src/main.rs

```diff
@@ -6,23 +6,77 @@
 use core::ptr;
 
 #[no_mangle]
-pub fn start() {
-    unsafe {
-        ptr::write_volatile(0xb8000 as *mut u16, 0x0248); // H
-        ptr::write_volatile(0xb8002 as *mut u16, 0x0265); // e
-        ptr::write_volatile(0xb8004 as *mut u16, 0x026c); // l
-        ptr::write_volatile(0xb8006 as *mut u16, 0x026c); // l
-        ptr::write_volatile(0xb8008 as *mut u16, 0x026f); // o
-        ptr::write_volatile(0xb800a as *mut u16, 0x022c); // ,
-        ptr::write_volatile(0xb800c as *mut u16, 0x0220); //
-        ptr::write_volatile(0xb800e as *mut u16, 0x0277); // w
-        ptr::write_volatile(0xb8010 as *mut u16, 0x026f); // o
-        ptr::write_volatile(0xb8012 as *mut u16, 0x0272); // r
-        ptr::write_volatile(0xb8014 as *mut u16, 0x026c); // l
-        ptr::write_volatile(0xb8016 as *mut u16, 0x0264); // d
-        ptr::write_volatile(0xb8018 as *mut u16, 0x0221); // !
+pub unsafe fn start() {
+    const NENTRIES: usize = 512;
+    const HUGE_PAGE_SIZE: u64 = 2 * 1024 * 1024; // 2 MiB
+
+    extern "C" {
+        static mut p2_table: [u64; NENTRIES];
+        static mut p3_table: [u64; NENTRIES];
+        static mut p4_table: [u64; NENTRIES];
     }
 
+    // Link up page tables
+    p4_table[0] = &p3_table[0] as *const _ as usize as u64 | 0b11;
+    p3_table[0] = &p2_table[0] as *const _ as usize as u64 | 0b11;
+
+    for (entry, i) in p2_table.iter_mut().zip(0..) {
+        *entry = i * HUGE_PAGE_SIZE | 0b10000011;
+    }
+
+    // move page table address to cr3
+    asm!("mov eax, p4_table
+          mov cr3, eax" :::: "intel");
```

Comment:
https://crates.io/crates/x86 would be neat to use here, but it also only supports 64-bit for now, IIRC.